### PR TITLE
Add support for internalize for external formats

### PR DESCRIPTION
**Pull request by hardikmaheshwari**
#### What type of PR is this?
It adds a new feature 
- adds internalize function
- the capability to deserialize externalFormats into model instance

#### What does this PR do / why is it needed ?
This PR adds 
- Definition of internalize functionality to deserialize external formats to model instances
- Organizes platform code for external formats - jsonSchema/flatdata/xsd to use this functionality

#### Which issue(s) this PR fixes:
It is linked with the issue - https://github.com/finos/legend-engine/issues/703 

#### Does this PR introduce a user-facing change?
No this adds a new functionality that cannot be currently used by users

### CODE REVIEW

File: legend-engine-external-shared-format-runtime/src/main/java/org/finos/legend/engine/external/shared/runtime/ExternalFormatExecutionExtension.java [hunk 1]
```diff
@@ -61,6 +63,10 @@ else if (executionNode instanceof UrlStreamExecutionNode)
             {
                 return executeUrlStream((UrlStreamExecutionNode) executionNode, pm, executionState);
             }
+            else if (executionNode instanceof VariableResolutionExecutionNode)
```
Comment (hardikmaheshwari) [inline legend-engine-external-shared-format-runtime/src/main/java/org/finos/legend/engine/external/shared/runtime/ExternalFormatExecutionExtension.java:66]: move this to core

File: legend-engine-external-shared-format-runtime/src/main/java/org/finos/legend/engine/external/shared/runtime/ExternalFormatExecutionExtension.java [hunk 2]
```diff
@@ -111,17 +127,31 @@ private Result executeUrlStream(UrlStreamExecutionNode node, MutableList<CommonP
         }
     }
 
+    private Result executeVariableResolution(VariableResolutionExecutionNode node, MutableList<CommonProfile> profiles, ExecutionState executionState)
```
Comment (hardikmaheshwari) [inline legend-engine-external-shared-format-runtime/src/main/java/org/finos/legend/engine/external/shared/runtime/ExternalFormatExecutionExtension.java:130]: move to core

File: legend-engine-pure-code-compiled-core/src/main/resources/core/pure/binding/executionPlan/executionPlan_diagram.pure [hunk 3]
```diff
@@ -27,18 +27,6 @@ Diagram meta::external::shared::format::executionPlan::diagram::ExternalSharedEx
         color=#FFFFCC,
         lineWidth=1.0)
 
-    TypeView cview_2(
```
Comment (hardikmaheshwari) [inline legend-engine-pure-code-compiled-core/src/main/resources/core/pure/binding/executionPlan/executionPlan_diagram.pure:30]: update executionplan_diagram to include varResolution node

File: legend-engine-pure-code-compiled-core/src/main/resources/core/pure/binding/executionPlan/executionPlan_print.pure [hunk 4]
```diff
@@ -54,6 +50,13 @@ function meta::external::shared::format::executionPlan::toString::printPlanNodeT
          $space + '  url = ' + $node.url + '\n' +
          $node->childrenToString($space+'  ', $extensions) + '\n' +
          $space + ')\n'
+      },
```
Comment (hardikmaheshwari) [inline legend-engine-pure-code-compiled-core/src/main/resources/core/pure/binding/executionPlan/executionPlan_print.pure:53]: move this up

File: legend-engine-pure-code-compiled-core/src/main/resources/core/pure/binding/executionPlan/javaRuntime/bindingLegendJavaPlatformBindingExtension.pure [hunk 5]
```diff
@@ -35,20 +35,14 @@ function meta::external::shared::format::executionPlan::platformBinding::legendJ
...
       updateGenerationContextForNode = {node: ExecutionNode[1], context: GenerationContext[1], path: String[1], extensions: Extension[*], debug: DebugContext[1] |
          $node->match([
-            {node:DataQualityExecutionNode[1] | 
-               print(if($debug.debug,|$debug.space+'('+$path+') prepare DataQualityExecutionNode\n', |''));
-               let childNodeInfo   = $context->nodeInfosForPath($path + '.n1')->toOne();
-               $context->addNodeInfo(^$childNodeInfo(path=$path));
-            },
             {node:ExternalFormatInternalizeExecutionNode[1] |
-               let externalFormatContract = $extensions.availableExternalFormats->cast(@ExternalFormatContract<Any>)->getExternalFormatContractForContentType($node.binding.contentType);
+               let externalFormatContract = $extensions.availableExternalFormats->cast(@ExternalFormatContract<Any>)->getExternalFormatContractForContentType($node.binding.contentType); //TODO: check this is cast needed
```
Comment (hardikmaheshwari) [inline legend-engine-pure-code-compiled-core/src/main/resources/core/pure/binding/executionPlan/javaRuntime/bindingLegendJavaPlatformBindingExtension.pure:45]: check this

File: legend-engine-pure-code-compiled-core/src/main/resources/core/pure/binding/executionPlan/javaRuntime/descriptor.pure [hunk 6]
```diff
@@ -12,19 +12,24 @@
...
 import meta::pure::executionPlan::engine::java::typeInfo::*;
 
+import meta::pure::executionPlan::platformBinding::legendJava::*;
+
 import meta::pure::extension::*;
 
 import meta::external::language::java::transform::*;
 
-Class meta::external::shared::format::executionPlan::engine::java::ExternalFormatJavaBindingDescriptor extends ExternalFormatPlatformBindingDescriptor
+Class meta::external::shared::format::executionPlan::engine::java::ExternalFormatLegendJavaPlatformBindingDescriptor extends LegendJavaPlatformBindingAdhocExtension
```
Comment (hardikmaheshwari) [inline legend-engine-pure-code-compiled-core/src/main/resources/core/pure/binding/executionPlan/javaRuntime/descriptor.pure:29]: review needed

File: legend-engine-pure-code-compiled-core/src/main/resources/core/pure/binding/executionPlan/model.pure [hunk 7]
```diff
@@ -29,29 +29,30 @@ import meta::pure::runtime::*;
...
-Class meta::external::shared::format::executionPlan::ExternalFormatExternalizeExecutionNode extends ExecutionNode
+Class meta::external::shared::format::executionPlan::UrlStreamExecutionNode extends ExecutionNode
 {
-   checked : Boolean[1];      
-   binding : Binding[1];
+   url : String[1];
 }
 
-Class meta::external::shared::format::executionPlan::ExternalFormatInternalizeExecutionNode extends ExecutionNode
+Class meta::external::shared::format::executionPlan::VariableResolutionExecutionNode extends meta::pure::executionPlan::ExecutionNode
```
Comment (hardikmaheshwari) [inline legend-engine-pure-code-compiled-core/src/main/resources/core/pure/binding/executionPlan/model.pure:53]: move this to core

File: legend-engine-pure-code-compiled-core/src/main/resources/core/pure/binding/executionPlan/model.pure [hunk 8]
```diff
@@ -67,9 +68,14 @@ Class meta::external::shared::format::executionPlan::UrlStreamExternalSource ext
    url : String[1];
 }
 
-Class meta::external::shared::format::executionPlan::ExecutionNodeExternalSource extends ExternalSource
+// ------------------------------------------------------------------------------------------------------------------------
+// URL
+// ------------------------------------------------------------------------------------------------------------------------
+
+//TODO: check this - move this to m3
+Class meta::pure::metamodel::type::Url
```
Comment (hardikmaheshwari) [inline legend-engine-pure-code-compiled-core/src/main/resources/core/pure/binding/executionPlan/model.pure:76]: review needed

File: legend-engine-pure-code-compiled-core/src/main/resources/core/pure/binding/protocols/pure/vX_X_X/extension/extension.pure [hunk 9]
```diff
@@ -27,8 +27,8 @@ function meta::protocols::pure::vX_X_X::external::shared::format::serializerExte
       transfers_executionPlan_transformNode =
          {mapping:Mapping[1], extensions:Extension[*] |
             [
-               d:meta::external::shared::format::executionPlan::DataQualityExecutionNode[1] | transformDataQualityNode($d, $mapping, $extensions),
-               u:meta::external::shared::format::executionPlan::UrlStreamExecutionNode[1]   | transformUrlStreamNode($u, $mapping, $extensions),
+               u:meta::external::shared::format::executionPlan::UrlStreamExecutionNode[1]                 | transformUrlStreamNode($u, $mapping, $extensions),
+               v:meta::external::shared::format::executionPlan::VariableResolutionExecutionNode[1]        | transformVariableResolutionExecutionNode($v, $mapping, $extensions),
```
Comment (hardikmaheshwari) [inline legend-engine-pure-code-compiled-core/src/main/resources/core/pure/binding/protocols/pure/vX_X_X/extension/extension.pure:31]: move this to core

File: legend-engine-pure-code-compiled-core/src/main/resources/core/pure/binding/protocols/pure/vX_X_X/models/executionPlan.pure [hunk 10]
```diff
@@ -57,4 +59,9 @@ Class meta::protocols::pure::vX_X_X::metamodel::external::shared::format::execut
 Class meta::protocols::pure::vX_X_X::metamodel::external::shared::format::executionPlan::UrlStreamExecutionNode extends ExecutionNode
 {
    url : String[1];
+}
+
+Class meta::protocols::pure::vX_X_X::metamodel::external::shared::format::executionPlan::VariableResolutionExecutionNode extends ExecutionNode
```
Comment (hardikmaheshwari) [inline legend-engine-pure-code-compiled-core/src/main/resources/core/pure/binding/protocols/pure/vX_X_X/models/executionPlan.pure:64]: move this to core

File: legend-engine-pure-code-compiled-core/src/main/resources/core/pure/executionPlan/legendJavaPlatformBinding/bindPlanToLegendJavaPlatform.pure [hunk 11]
```diff
@@ -32,14 +32,16 @@ function meta::pure::executionPlan::platformBinding::legendJava::bindPlanToLegen
    let legendJavaConfig   = $platformBindingConfig->cast(@LegendJavaPlatformBindingConfig);
    let generationContext  = $plan->prepareGenerationContext($legendJavaConfig, $newExtensions, $debug);
 
+   let baseProject        = $generationContext->generateTypes($debug);
```
Comment (hardikmaheshwari) [inline legend-engine-pure-code-compiled-core/src/main/resources/core/pure/executionPlan/legendJavaPlatformBinding/bindPlanToLegendJavaPlatform.pure:35]: review needed

File: legend-engine-pure-code-compiled-core/src/main/resources/core/pure/graphFetch/graphExtension.pure [hunk 12]
```diff
@@ -448,28 +469,31 @@ function <<access.private>> meta::pure::graphFetch::enrichSourceTreeNodeForPrope
...
 
-   if($isPropertyTemporalMilestoned,
-      | ^$withQPsRecorded(subTrees=$withQPsRecorded.subTrees->map(st| $st->switchToMilestoneProperties($tgtPgft.parameters->evaluateAndDeactivate()))),
-      | $withQPsRecorded
+         if($isPropertyTemporalMilestoned,
+            | ^$withQPsRecorded(subTrees=$withQPsRecorded.subTrees->map(st| $st->switchToMilestoneProperties($tgtPgft.parameters->evaluateAndDeactivate()))),
+            | $withQPsRecorded
+         );
+      },
+      | $srcNode  //TOD: check this
```
Comment (hardikmaheshwari) [inline legend-engine-pure-code-compiled-core/src/main/resources/core/pure/graphFetch/graphExtension.pure:496]: check this

File: legend-engine-pure-code-compiled-core/src/main/resources/core/pure/graphFetch/graphExtension.pure [hunk 13]
```diff
@@ -539,6 +563,13 @@ function meta::pure::graphFetch::propertyTreeToGraphFetchTree(pTree:PropertyPath
    $root->addPropertyGraphFetchTrees($pTree, $ownerClass)->cast(@RootGraphFetchTree<Any>);
 }
 
+function meta::pure::graphFetch::propertyGraphFetchTreeToRootGraphFetchTree(pTree:PropertyGraphFetchTree[1]): RootGraphFetchTree<Any>[1]
```
Comment (hardikmaheshwari) [inline legend-engine-pure-code-compiled-core/src/main/resources/core/pure/graphFetch/graphExtension.pure:566]: review needed

File: legend-engine-pure-code-compiled-core/src/main/resources/core/pure/graphFetch/graphFetch_routing.pure [hunk 14]
```diff
@@ -254,12 +257,14 @@ function <<access.private>> meta::pure::graphFetch::routing::checkForConstraints
...
 {
-  let properties = $node.subTrees->filter(n | $n->instanceOf(PropertyGraphFetchTree))->map(n | $n->cast(@PropertyGraphFetchTree).property);
-  let pathsForConstraint = pathsForConstraintFunctions($class, $f)->removeDuplicates();
+  let pathsForConstraint        = pathsForConstraintFunctions($class, $f)->removeDuplicates();
+  let qualifiedPropertyPaths    = $pathsForConstraint->filter(path| $path.values->exists(x| $x.property->instanceOf(QualifiedProperty)));
+  let inlinedPropertyTree       = $pathsForConstraint->buildPropertyTree()->inlineQualifiedPropertyNodes();
+  let inlinedGraphTree          = $inlinedPropertyTree->propertyTreeToGraphFetchTree($class);
 
-  $pathsForConstraint->size() > $properties->size();
+  !canEvaluateForTree($inlinedGraphTree, $gfTree->propertyGraphFetchTreeToRootGraphFetchTree());
```
Comment (hardikmaheshwari) [inline legend-engine-pure-code-compiled-core/src/main/resources/core/pure/graphFetch/graphFetch_routing.pure:267]: review needed

File: legend-engine-pure-code-compiled-core/src/main/resources/core/store/platform/executionPlan/executionPlan_generation.pure [hunk 15]
```diff
@@ -13,62 +13,110 @@
...
-
+import meta::pure::graphFetch::*;
+import meta::pure::router::clustering::*;
 import meta::pure::runtime::*;
+import meta::pure::store::*;
 import meta::pure::store::platform::executionPlan::generation::*;
+import meta::pure::store::platform::metamodel::*;
 
-Class meta::pure::store::platform::executionPlan::generation::PureFunctionToProcessFunctionPair extends Pair<meta::pure::metamodel::function::Function<Any>, meta::pure::metamodel::function::Function<{FunctionExpression[1], Runtime[0..1], Map<String, List<Any>>[1], Boolean[0..1], ExecutionContext[1], Extension[*], DebugContext[1] -> ExecutionNode[1]}>>
+Class meta::pure::store::platform::executionPlan::generation::PlatformPlanGenerationState
```
Comment (hardikmaheshwari) [inline legend-engine-pure-code-compiled-core/src/main/resources/core/store/platform/executionPlan/executionPlan_generation.pure:24]: review needed

File: legend-engine-xt-flatdata-pure/src/main/resources/core_external_format_flatdata/externalFormatContract.pure [hunk 16]
```diff
@@ -45,6 +48,9 @@ function meta::external::format::flatdata::extension::flatDataFormatExtension():
    ^Extension
    (
       type                     = 'External Format - FlatData',
-      availableExternalFormats = flatDataFormatContract()
+      availableExternalFormats = flatDataFormatContract(),
+
+      //TODO: Move this out along with Java Binding
+      availablePlatformBindings = externalFormatLegendJavaPlatformBinding(flatDataJavaBindingDescriptor())
```
Comment (hardikmaheshwari) [inline legend-engine-xt-flatdata-pure/src/main/resources/core_external_format_flatdata/externalFormatContract.pure:54]: review needed

File: legend-engine-xt-json-pure/src/main/resources/core_external_format_json/executionPlan/javaRuntime/internalize.pure [hunk 17]
```diff
@@ -16,6 +16,7 @@ import meta::external::format::json::binding::validation::*;
 import meta::external::format::json::executionPlan::model::*;
 import meta::external::format::json::executionPlan::engine::java::*;
 import meta::external::format::json::executionPlan::engine::java::internalize::*;
+import meta::external::format::json::executionPlan::engine::java::internalize::jsonReader::*;
```
Comment (hardikmaheshwari) [inline legend-engine-xt-json-pure/src/main/resources/core_external_format_json/executionPlan/javaRuntime/internalize.pure:19]: remove this

File: legend-engine-xt-relationalStore-pure/src/main/resources/core_relational/relational/milestoning/tests/testBusinessDateMilestoning.pure [hunk 18]
```diff
@@ -51,7 +51,7 @@ function <<test.Test>> meta::relational::tests::milestoning::businessdate::testC
...
+   assertSameSQL('select \'CST\' as "CONSTRAINT_ID", \'Error\' as "ENFORCEMENT_LEVEL", \'Ensure parent (this) milestoning context propagated through project\' as "MESSAGE", "root".id as "id", "root".name as "name" from ProductTable as "root" where not (not (not exists(select 1 from SystemTable as "systemtable_0" left outer join (select "systemdescriptiontable_1".systemName as systemName, "systemdescriptiontable_1".description as description from SystemDescriptionTable as "systemdescriptiontable_1" where "systemdescriptiontable_1".from_z <= \'2015-10-16\' and "systemdescriptiontable_1".thru_z > \'2015-10-16\') as "systemdescriptiontable_0" on ("systemtable_0".name = "systemdescriptiontable_0".systemName) where "root".referenceSystemName = "systemtable_0".name))) and "root".from_z <= \'2015-10-16\' and "root".thru_z > \'2015-10-16\'', $result);
```
Comment (hardikmaheshwari) [inline legend-engine-xt-relationalStore-pure/src/main/resources/core_relational/relational/milestoning/tests/testBusinessDateMilestoning.pure:54]: revert this file

File: legend-engine-xt-relationalStore-pure/src/main/resources/core_relational/relational/pureToSQLQuery/pureToSQLQuery.pure [hunk 19]
```diff
@@ -2986,7 +2986,7 @@ function meta::relational::functions::pureToSqlQuery::processDynaFunction(f:Func
    let res = ^$operation(select = $state.inFilter->if(
                            | let unSupportedBiTemporalProcessing = $mergedSQL.filteringOperation->isEmpty() && $operation.milestoningContext->isNotEmpty() && $operation.milestoningContext->toOne()->isBiTemporalMilestoningTopLevelAnd($f.func);
                              if($unSupportedBiTemporalProcessing ,| $mergedSQL
-                                                                 ,| ^$mergedSQL(filteringOperation = newDynaFunction($f.func.functionName->toOne(), $mergedSQL.filteringOperation)));,
+                                                                 ,| ^$mergedSQL(filteringOperation = newDynaFunction($f.func.functionName->toOne(), if($parameterSQL->isEmpty(), | [], |$mergedSQL.filteringOperation))));,
```
Comment (hardikmaheshwari) [inline legend-engine-xt-relationalStore-pure/src/main/resources/core_relational/relational/pureToSQLQuery/pureToSQLQuery.pure:2989]: revert this file

**Review by gs-ssh16 — APPROVED** [review]

**Pull request merged**